cmd/thanos/store.go

```diff
@@ -124,17 +131,18 @@ func runStore(
 			return errors.Wrap(err, "create object storage store")
 		}
 
-		begin := time.Now()
-		level.Debug(logger).Log("msg", "initializing bucket store")
-		if err := bs.InitialSync(context.Background()); err != nil {
-			return errors.Wrap(err, "bucket store initial sync")
-		}
-		level.Debug(logger).Log("msg", "bucket store ready", "init_duration", time.Since(begin).String())
-
 		ctx, cancel := context.WithCancel(context.Background())
 		g.Add(func() error {
 			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 
+			begin := time.Now()
+			level.Debug(logger).Log("msg", "initializing bucket store")
+			if err := bs.InitialSync(ctx); err != nil {
+				return errors.Wrap(err, "bucket store initial sync")
+			}
+			level.Debug(logger).Log("msg", "bucket store ready", "init_duration", time.Since(begin).String())
+			readinessProber.SetReady()
+
```

Comment:
@bwplotka 
I'm not completely sure if this is correct. As I wrote in the PR description I had to move the initialization of cache to the actor so the HTTP server can start up otherwise it was blocking it. I'm setting the store as ready after the cache is inicialized but in setup not honoring the readiness probe the store could be already queried.

Reply:
Yes, this is very tricky part. I don't think we should NOT set readiness until we got ready to serve gRPC request - even though we cannot serve metrics there. Not sure yet how to solve it.

I think the idea is to move to faster store startup by moving certain things to compactor (see https://github.com/improbable-eng/thanos/issues/942). Having that said, can we rever this for now? and initial sync BEFORE setting Ready? Let's discuss and improve this in later PR (: For now moving to probe and ensuring all users are using that is our goal, right?

Reply:
Agree, the issue now is that you have to add _some_ initiallBackOff for the liveness probe to make sure it does not get restarted before.

But as I said myself I knot this is tricky part so I'll be happy to rollback the change as you suggested.